jena-arq/src/main/java/org/apache/jena/sparql/core/DynamicDatasets.java

```diff
@@ -116,7 +117,30 @@ public static class DynamicDatasetGraph extends DatasetGraphReadOnly implements
 
         public DynamicDatasetGraph(DatasetGraph viewDSG, DatasetGraph baseDSG) {
             super(viewDSG, baseDSG.getContext().copy());
-            this.projected = baseDSG;
+            this.projected = Objects.requireNonNull(baseDSG);
+        }
+
+        /** Returns the original dataset graph this DynamicDatasetGraph is based on. Never null. */
+        public DatasetGraph getProjected() {
+            return projected;
+        }
+
+        /** Returns the collection of default graphs this DynamicDatasetGraph is based on. Null if unknown.
+         * @implNote This method returns the context's value for {@link ARQConstants#symDatasetDefaultGraphs}. */
+        public Collection<Node> getDefaultGraphs() {
+            return getContext().get(ARQConstants.symDatasetDefaultGraphs);
+        }
+
+        /** Returns the collection of named graphs this DynamicDatasetGraph is based on. Null if unknown.
+         * @implNote This method returns the context's value for {@link ARQConstants#symDatasetNamedGraphs}. */
+        public Collection<Node> getNamedGraphs() {
+            return getContext().get(ARQConstants.symDatasetNamedGraphs);
+        }
+
+        /** True iff {@link #getDefaultGraphs()} and {@link #getNamedGraphs()} both
+         * return non-null results. */
+        public boolean hasDatasetDescription() {
```

Comment:
The rationale is that this is the test for whether the DynamicDataset in unwrappable. 
Because the getters depend on the context attributes (they could also be stored as fields) it is possible to unset them from the outside - in this case the DynamicDataset getters return null which means that it is not un-wrappable.

Reply:
To access this method, the object has to be cast. The class is public so that it can be instance-tested.

The dataset context is a copy to freeze it.